### PR TITLE
New Datasource: `azurerm_hybrid_compute_machine`

### PR DESCRIPTION
**Pull request by liuwuliuyun**
New Datasource: `azurerm_hybrid_compute_machine`

Swagger: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/hybridcompute/resource-manager/Microsoft.HybridCompute/stable/2022-11-10/HybridCompute.json

API Version: 2022-11-10

Note: In the acctest of this data source, I use an Azure VM to fake an on-premises machine, so it needs special configuration which is in the install_arc.sh.tmpl file. And the service principal running this test should at least have [Virtual Machine Contributor](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#virtual-machine-contributor) role.

### CODE REVIEW

**Comment by liuwuliuyun** [comment]
Working on the CI promoted problems.

**Comment by liuwuliuyun** [comment]
Testing evidence:

GOROOT=C:\Users\yunliu1\go\go1.19.3 #gosetup
GOPATH=C:\Users\yunliu1\go #gosetup
C:\Users\yunliu1\go\go1.19.3\bin\go.exe test -c -o C:\Users\yunliu1\AppData\Local\Temp\GoLand\___TestAccHybridComputeMachine_basic_in_github_com_hashicorp_terraform_provider_azurerm_internal_services_hybridcompute.test.exe github.com/hashicorp/terraform-provider-azurerm/internal/services/hybridcompute #gosetup
C:\Users\yunliu1\go\go1.19.3\bin\go.exe tool test2json -t C:\Users\yunliu1\AppData\Local\Temp\GoLand\___TestAccHybridComputeMachine_basic_in_github_com_hashicorp_terraform_provider_azurerm_internal_services_hybridcompute.test.exe -test.v -test.paniconexit0 -test.run ^\QTestAccHybridComputeMachine_basic\E$ #gosetup
=== RUN   TestAccHybridComputeMachine_basic
=== PAUSE TestAccHybridComputeMachine_basic
=== CONT  TestAccHybridComputeMachine_basic
--- PASS: TestAccHybridComputeMachine_basic (559.64s)
PASS


Process finished with the exit code 0

File: internal/services/hybridcompute/hybrid_compute_machine_data_source.go [hunk 1]
```diff
@@ -0,0 +1,1111 @@
...
+func (r HybridComputeMachineDataSource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return machines.ValidateMachineID
+}
+
+func (r HybridComputeMachineDataSource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ForceNew: true,
```
[the remark this answers is not in the document]
Comment (liuwuliuyun) [reply internal/services/hybridcompute/hybrid_compute_machine_data_source.go:154]: Thanks @ms-zhenhua for the review, I will remove the force new for these two properties.

File: internal/services/hybridcompute/hybrid_compute_machine_data_source_test.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  features {
+    resource_group {
+      prevent_deletion_if_contains_resources = false
+    }
+  }
+}
+
+data "azurerm_client_config" "current" {}
+
+# note: real-life usage perfer random_uuid resource in registry.terraform.io/hashicorp/random
```
Comment (manicminer) [inline internal/services/hybridcompute/hybrid_compute_machine_data_source_test.go:42]:
```suggestion
# note: real-life usage prefer random_uuid resource in registry.terraform.io/hashicorp/random
```

File: internal/services/hybridcompute/hybrid_compute_machine_data_source_test.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    name                       = "SSH"
+    priority                   = 1001
+    direction                  = "Inbound"
+    access                     = "Allow"
+    protocol                   = "Tcp"
+    source_port_range          = "*"
+    destination_port_range     = "22"
+    source_address_prefix      = "*"
+    destination_address_prefix = "*"
+  }
```
Comment (manicminer) [inline internal/services/hybridcompute/hybrid_compute_machine_data_source_test.go:94]: Does this need to be open to the whole Internet? If it does, then we should be generating a random login password.

File: internal/services/hybridcompute/scripts/install_arc.sh.tmpl [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+
+# Disable the Azure VM Guest Agent.
+current_hostname=$(hostname);
+sudo service walinuxagent stop;
+sudo waagent -deprovision -force;
+sudo rm -rf /var/lib/waagent;
+sudo hostnamectl set-hostname $current_hostname;
```
Comment (manicminer) [inline internal/services/hybridcompute/scripts/install_arc.sh.tmpl:8]: Semicolon termination not necessary
```suggestion
current_hostname=$(hostname)
sudo service walinuxagent stop
sudo waagent -deprovision -force
sudo rm -rf /var/lib/waagent
sudo hostnamectl set-hostname $current_hostname
```

File: internal/services/hybridcompute/scripts/install_arc.sh.tmpl [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+export cloud="AzureCloud";
+output=$(wget https://aka.ms/azcmagent -O ~/install_linux_azcmagent.sh 2>&1);
+if [ $? != 0 ];
+then wget -qO- --method=PUT --body-data="{\"subscriptionId\":\"$subscriptionId\",\"resourceGroup\":\"$resourceGroup\",\"tenantId\":\"$tenantId\",\"location\":\"$location\",\"correlationId\":\"$correlationId\",\"authType\":\"$authType\",\"operation\":\"onboarding\",\"messageType\":\"DownloadScriptFailed\",\"message\":\"$output\"}" "https://gbl.his.arc.azure.com/log" &> /dev/null || true;
+fi;
+echo "$output";
+sudo chmod +x ~/install_linux_azcmagent.sh;
+cd ~;
+./install_linux_azcmagent.sh;
+sudo azcmagent connect --service-principal-id "$servicePrincipalClientId" --service-principal-secret "$servicePrincipalSecret" --resource-group "$resourceGroup" --tenant-id "$tenantId" --location "$location" --subscription-id "$subscriptionId" --cloud "$cloud" --correlation-id "$correlationId";
```
Comment (manicminer) [inline internal/services/hybridcompute/scripts/install_arc.sh.tmpl:33]: As above, also suggest quoting some of these values as they may contain special chars, and consider using a heredoc & splitting newines for readability.

Do the exported variables need to be exported? If not, making them local variables improves readability (I have done this in the suggestion).

```suggestion
servicePrincipalClientId="${client_id}"
servicePrincipalSecret="${client_secret}"
cloud="AzureCloud"
tenantId="${tenant_id}"
subscriptionId="${subscription_id}"
resourceGroup="${resource_group_name}"
location="${location}"
authType="principal"
correlationId="${uuid}"

output=$(wget https://aka.ms/azcmagent -O ~/install_linux_azcmagent.sh 2>&1)

if [ $? != 0 ]; then
  read bodyData <<EOF
  {
    "operation": "onboarding",
    "messageType": "DownloadScriptFailed",
    "tenantId": "${tenantId}",
    "subscriptionId": "${subscriptionId}",
    "resourceGroup": "${resourceGroup}",
    "location": "${location}",
    "correlationId": "${correlationId}",
    "authType": "${authType}",
    "message": "${output}"
  }
EOF

  wget -qO- --method=PUT --body-data="${bodyData}" "https://gbl.his.arc.azure.com/log" &> /dev/null || true
fi

echo "${output}"

sudo chmod +x ~/install_linux_azcmagent.sh
cd ~
./install_linux_azcmagent.sh

sudo azcmagent connect \
  --service-principal-id "${servicePrincipalClientId}" \
  --service-principal-secret "${servicePrincipalSecret}" \
  --cloud "${cloud}" \
  --tenant-id "${tenantId}" \
  --subscription-id "${subscriptionId}" \
  --resource-group "${resourceGroup}" \
  --location "${location}" \
  --correlation-id "${correlationId}"
```

File: internal/services/hybridcompute/hybrid_compute_machine_data_source_test.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+func TestAccHybridComputeMachine_basic(t *testing.T) {
+	data := acceptance.BuildTestData(t, "data.azurerm_hybrid_compute_machine", "test")
+	d := HybridComputeMachineDataSource{}
+	clientSecret := os.Getenv("ARM_CLIENT_SECRET")
+	randomUUID, _ := uuid.GenerateUUID()
+	data.DataSourceTest(t, []acceptance.TestStep{
+		{
+			Config: d.basic(data, clientSecret, randomUUID),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).Key("agent_configuration.#").HasValue("1"),
```
Comment (manicminer) [inline internal/services/hybridcompute/hybrid_compute_machine_data_source_test.go:24]: Are there any other checks we can reliably add here, given the complexity of this data source?

File: internal/services/hybridcompute/hybrid_compute_machine_data_source.go [hunk 7]
```diff
@@ -0,0 +1,1103 @@
...
+
+	output := OSProfileModel{}
+
+	if input.ComputerName != nil {
+		output.ComputerName = *input.ComputerName
+	}
+
+	linuxConfigurationValue := flattenOSProfileLinuxConfigurationModel(input.LinuxConfiguration)
+
+	output.LinuxConfiguration = linuxConfigurationValue
```
Comment (manicminer) [inline internal/services/hybridcompute/hybrid_compute_machine_data_source.go:1009]: Readability
```suggestion
	linuxConfigurationValue := flattenOSProfileLinuxConfigurationModel(input.LinuxConfiguration)
	output.LinuxConfiguration = linuxConfigurationValue
```

File: internal/services/hybridcompute/hybrid_compute_machine_data_source.go [hunk 8]
```diff
@@ -0,0 +1,1103 @@
...
+
+	windowsConfigurationValue := flattenOSProfileWindowsConfigurationModel(input.WindowsConfiguration)
+
+	output.WindowsConfiguration = windowsConfigurationValue
```
Comment (manicminer) [inline internal/services/hybridcompute/hybrid_compute_machine_data_source.go:1013]:
```suggestion
	windowsConfigurationValue := flattenOSProfileWindowsConfigurationModel(input.WindowsConfiguration)
	output.WindowsConfiguration = windowsConfigurationValue
```

File: internal/services/hybridcompute/hybrid_compute_machine_data_source.go [hunk 9]
```diff
@@ -0,0 +1,1103 @@
...
+	var outputList []OSProfileLinuxConfigurationModel
+	if input == nil {
+		return outputList
+	}
+
+	output := OSProfileLinuxConfigurationModel{}
+
+	patchSettingsValue := flattenPatchSettingsModel(input.PatchSettings)
+
+	output.PatchSettings = patchSettingsValue
```
Comment (manicminer) [inline internal/services/hybridcompute/hybrid_compute_machine_data_source.go:1028]:
```suggestion
	output := OSProfileLinuxConfigurationModel{}
	patchSettingsValue := flattenPatchSettingsModel(input.PatchSettings)
	output.PatchSettings = patchSettingsValue
```

File: internal/services/hybridcompute/hybrid_compute_machine_data_source.go [hunk 10]
```diff
@@ -0,0 +1,1103 @@
...
+	var outputList []OSProfileWindowsConfigurationModel
+	if input == nil {
+		return outputList
+	}
+
+	output := OSProfileWindowsConfigurationModel{}
+
+	patchSettingsValue := flattenPatchSettingsModel(input.PatchSettings)
+
+	output.PatchSettings = patchSettingsValue
```
Comment (manicminer) [inline internal/services/hybridcompute/hybrid_compute_machine_data_source.go:1062]:
```suggestion
	output := OSProfileWindowsConfigurationModel{}
	patchSettingsValue := flattenPatchSettingsModel(input.PatchSettings)
	output.PatchSettings = patchSettingsValue
```

File: internal/services/hybridcompute/hybrid_compute_machine_data_source.go [hunk 11]
```diff
@@ -0,0 +1,1103 @@
...
+	var outputList []ServiceStatusesModel
+	if input == nil {
+		return outputList
+	}
+
+	output := ServiceStatusesModel{}
+
+	extensionServiceValue := flattenServiceStatusModel(input.ExtensionService)
+
+	output.ExtensionService = extensionServiceValue
```
Comment (manicminer) [inline internal/services/hybridcompute/hybrid_compute_machine_data_source.go:1077]:
```suggestion
	extensionServiceValue := flattenServiceStatusModel(input.ExtensionService)
	output.ExtensionService = extensionServiceValue
```

File: internal/services/hybridcompute/hybrid_compute_machine_data_source.go [hunk 12]
```diff
@@ -0,0 +1,1103 @@
...
+
+	guestConfigurationServiceValue := flattenServiceStatusModel(input.GuestConfigurationService)
+
+	output.GuestConfigurationService = guestConfigurationServiceValue
```
Comment (manicminer) [inline internal/services/hybridcompute/hybrid_compute_machine_data_source.go:1081]:
```suggestion
	guestConfigurationServiceValue := flattenServiceStatusModel(input.GuestConfigurationService)
	output.GuestConfigurationService = guestConfigurationServiceValue
```

**Review by manicminer — CHANGES_REQUESTED** [review]
Hi @liuwuliuyun, thanks for this PR. On the whole, this is looking great. I have some questions about the schema, and have made some suggestions mostly around readability. If you can take a look at these, I'll be happy to take another look and I look forward to merging this. Thanks!

File: internal/services/hybridcompute/hybrid_compute_machine_data_source_test.go [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+  connection {
+    type     = "ssh"
+    host     = azurerm_public_ip.test.ip_address
+    user     = "adminuser"
+    password = "P@$$w0rd1234!"
+  }
+
+  provisioner "file" {
+    content = templatefile("scripts/install_arc.sh.tmpl", {
```
Comment (manicminer) [inline internal/services/hybridcompute/hybrid_compute_machine_data_source_test.go:143]: The suggested file extension here is `.tftpl`
```suggestion
    content = templatefile("scripts/install_arc.sh.tftpl", {
```

**Comment by liuwuliuyun** [comment]
Thanks for the review, I will take a look today.

File: internal/services/hybridcompute/hybrid_compute_machine_data_source.go [hunk 14]
```diff
@@ -0,0 +1,1103 @@
...
+)
+
+type HybridComputeMachineModel struct {
+	Name                       string                              `tfschema:"name"`
+	ResourceGroupName          string                              `tfschema:"resource_group_name"`
+	AgentConfiguration         []AgentConfigurationModel           `tfschema:"agent_configuration"`
+	ClientPublicKey            string                              `tfschema:"client_public_key"`
+	CloudMetadata              []CloudMetadataModel                `tfschema:"cloud_metadata"`
+	DetectedProperties         map[string]string                   `tfschema:"detected_properties"`
+	Extensions                 []MachineExtensionInstanceViewModel `tfschema:"extensions"`
```
[the remark this answers is not in the document]
Comment (liuwuliuyun) [reply internal/services/hybridcompute/hybrid_compute_machine_data_source.go:27]: I could not find MachineExtension type, so this seems to be a sdk generation problem with hashicorp/go-azure-sdk.

[the remark this answers is not in the document]
Comment (liuwuliuyun) [reply internal/services/hybridcompute/hybrid_compute_machine_data_source.go:27]: But from the official [doc](https://learn.microsoft.com/en-us/rest/api/hybridcompute/machines/create-or-update?tabs=HTTP), this field is deprecated, so I will remove this field from the property list.

File: internal/services/hybridcompute/hybrid_compute_machine_data_source.go [hunk 15]
```diff
@@ -0,0 +1,1103 @@
...
+	Location                   string                              `tfschema:"location"`
+	LocationData               []LocationDataModel                 `tfschema:"location_data"`
+	MssqlDiscovered            string                              `tfschema:"mssql_discovered"`
```
[the remark this answers is not in the document]
Comment (liuwuliuyun) [reply internal/services/hybridcompute/hybrid_compute_machine_data_source.go:30]: Changed to bool

File: internal/services/hybridcompute/hybrid_compute_machine_data_source.go [hunk 16]
```diff
@@ -0,0 +1,1103 @@
...
+	if input.Level != nil {
+		output.Level = *input.Level
+	}
+
+	if input.Message != nil {
+		output.Message = *input.Message
+	}
+
+	if input.Time != nil {
+		output.Time = *input.Time
```
[the remark this answers is not in the document]
Comment (liuwuliuyun) [reply internal/services/hybridcompute/hybrid_compute_machine_data_source.go]: Deprecated as above, remove from property list.

File: website/docs/d/hybrid_compute_machine.html.markdown [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+
+* `code` - The status code.
+
+* `display_status` - The short localizable label for the status.
+
+* `level` - The level code.
+
+* `message` - The detailed status message, including for alerts and error messages.
+
+* `time` - The time of the status.
```
[the remark this answers is not in the document]
Comment (liuwuliuyun) [reply website/docs/d/hybrid_compute_machine.html.markdown:252]: Same reason above, remove from document..

File: internal/services/hybridcompute/hybrid_compute_machine_data_source_test.go (hunk 13 above)
Comment (liuwuliuyun) [reply internal/services/hybridcompute/hybrid_compute_machine_data_source_test.go:143]: Changed to suggested extension name.

**Comment by liuwuliuyun** [comment]
Hi @manicminer, I have changed the code according to your review comment. I would be very appreciated if you could take another look. :)

File: internal/services/hybridcompute/scripts/install_arc.sh.tftpl [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+servicePrincipalClientId="${client_id}"
+servicePrincipalSecret="${client_secret}"
+cloud="AzureCloud"
+tenantId="${tenant_id}"
+subscriptionId="${subscription_id}"
+resourceGroup="${resource_group_name}"
+location="${location}"
+authType="principal"
+correlationId="${uuid}"
+output=$(wget https://aka.ms/azcmagent -O ~/install_linux_azcmagent.sh 2>&1)
```
Comment (manicminer) [inline internal/services/hybridcompute/scripts/install_arc.sh.tftpl:25]: Quotes here for shell safety
```suggestion
output="$(wget https://aka.ms/azcmagent -O ~/install_linux_azcmagent.sh 2>&1)"
```

**Review by manicminer — APPROVED** [review]
Thanks for making the changes @liuwuliuyun, this LGTM 👍

**Comment by manicminer** [comment]
New service, test results

Screenshot 2023-02-13 at 14 07 49

**Pull request merged**